app/models/pglogical_subscription.rb

```diff
@@ -63,58 +69,95 @@ def self.save_all!(subscription_list)
+  # argument is nil. Afterwards the evm-failover-monitor service is run as
+  # a queued operation.
+  #
   def self.delete_all(list = nil)
     (list.nil? ? find(:all) : list)&.each { |sub| sub.delete(false) }
     EvmDatabase.restart_failover_monitor_service_queue
     nil
   end
 
+  # Wraps the Pg::LogicalReplicationClient#disable_subscription
+  # method, then checks the result state.
+  #
   def disable
     pglogical.disable_subscription(id).check
   end
 
+  # Wraps the Pg::LogicalReplicationClient#enable_subscription
+  # method, then checks the result state.
+  #
   def enable
     pglogical.enable_subscription(id).check
   end
 
+  # The postgres logical replication client. Creates a new
+  # PG::LogicalReplication::Client if the +refresh+ argument is set to true.
+  # By default the +refresh+ argument is set to false.
+  #
   def self.pglogical(refresh = false)
     @pglogical = nil if refresh
     @pglogical ||= PG::LogicalReplication::Client.new(connection.raw_connection)
   end
 
+  # Instance method wrapper around the +pglogical+ singleton method.
+  #
   def pglogical(refresh = false)
     self.class.pglogical(refresh)
   end
 
+  # Validate the +new_connection_params+ hash using the MiqRegionRemote model.
+  # The keys it looks for are password, host, port, user and dbname.
+  #
+  # If the params[:password] is blank it wil parse it out of the
+  # subscription DSN.
+  #
   def validate(new_connection_params = {})
-    find_password if new_connection_params['password'].blank?
-    connection_hash = attributes.merge(new_connection_params.delete_blanks)
-    MiqRegionRemote.validate_connection_settings(connection_hash['host'],
-                                                 connection_hash['port'],
-                                                 connection_hash['user'],
-                                                 connection_hash['password'],
-                                                 connection_hash['dbname'])
-  end
-
+    params = new_connection_params.symbolize_keys
+    find_password if params[:password].blank?
+    connection_hash = attributes.merge(params.delete_blanks)
+
+    MiqRegionRemote.validate_connection_settings(
+      connection_hash[:host],
+      connection_hash[:port],
+      connection_hash[:user],
+      connection_hash[:password],
+      connection_hash[:dbname]
+    )
+  end
+
+  # Essentially a wrapper around our custom +xlog_location_diff+ method, defined
+  # in the ar_dba.rb file. However, this will instead log an error and return nil
+  # if the current status isn't "replicating".
+  #
   def backlog
     if status != "replicating"
       _log.error("Is `#{dbname}` running on host `#{host}` and accepting TCP/IP connections on port #{port} ?")
       return nil
     end
     begin
-      connection.xlog_location_diff(remote_region_lsn, subscription_attributes["remote_replication_lsn"])
+      connection.xlog_location_diff(remote_region_lsn, remote_replication_lsn)
```

Comment:
Right, I kept it as a property. Currently that information is deleted at:

https://github.com/ManageIQ/manageiq/blob/master/app/models/pglogical_subscription.rb#L131

But then we seem to need it here, which requires another separate call:

https://github.com/ManageIQ/manageiq/blob/master/app/models/pglogical_subscription.rb#L111

So, rather than make an extra call to get at information we deleted earlier, I kept it around.

Reply:
The problem is we need the information as it is currently. This value changes every time a byte is replicated over the wire. If it is cached then we don't have an accurate view of the backlog.

Reply:
(this comment was meant to be here)

Ok, understood, but that is precisely the sort of thing that needs commenting. Without being intimately familiar with pglogical, that is non-obvious.

Reply:
I agree that this maybe requires a bit more explanation so a comment here would make sense, but my point about comments in my review is that this proposed explanation is very different than some of the comments on the simpler methods.